test/concurrent_async_map_test.dart

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:test/test.dart';
+
+import 'package:stream_transform/stream_transform.dart';
+
+void main() {
+  var streamTypes = {
+    'single subscription': () => new StreamController(),
+    'broadcast': () => new StreamController.broadcast()
+  };
+  StreamController streamController;
+  List emittedValues;
+  bool valuesCanceled;
+  bool isDone;
+  List errors;
+  Stream transformed;
+  StreamSubscription subscription;
+
+  List<Completer> finishWork;
+  List<dynamic> values;
+
+  Future convert(dynamic value) {
+    values.add(value);
+    var completer = new Completer();
+    finishWork.add(completer);
+    return completer.future;
+  }
+
+  for (var streamType in streamTypes.keys) {
+    group('concurrentAsyncMap for stream type: [$streamType]', () {
+      setUp(() {
+        valuesCanceled = false;
+        streamController = streamTypes[streamType]()
+          ..onCancel = () {
+            valuesCanceled = true;
+          };
+        emittedValues = [];
+        errors = [];
+        isDone = false;
+        finishWork = [];
+        values = [];
+        transformed =
+            streamController.stream.transform(concurrentAsyncMap(convert));
+        subscription = transformed
+            .listen(emittedValues.add, onError: errors.add, onDone: () {
+          isDone = true;
+        });
+      });
+
+      test('does not emit before convert finishes', () async {
+        streamController.add(1);
+        await new Future(() {});
+        expect(emittedValues, isEmpty);
+        expect(values, [1]);
+        finishWork.first.complete(1);
+        await new Future(() {});
+        expect(emittedValues, [1]);
+      });
+
+      test('allows calls to convert before the last one finished', () async {
+        streamController.add(1);
+        streamController.add(2);
+        streamController.add(3);
+        await new Future(() {});
+        expect(values, [1, 2, 3]);
```

Comment:
How is this test ensuring the convert can be called before the last one? Is it implicit that they are completed before a single pass through the event loop?

Reply:
items aren't added to `values` until the convert callback is called, and the convert callback doesn't finish until you call `finishWork[x].complete()` so this is indicating that 3 values have been passed to `convert` before any value has finished.

Reply:
Ahh I see now.